### ISSUE TITLE
Unable to use any mapping containing 'CTRL+\' in apps

### ISSUE THREAD

**Issue by ojroques** [microsoft/terminal#8458]
# Environment

```none
Windows build number: 10.0.19041.0
Windows Terminal version (if applicable): Windows Terminal Preview v1.5.3242.0
```

# Steps to reproduce
1. Try to use a mapping containing `CTRL+\`: for instance `CTRL+\ CTRL+n` in Neovim to quit the built-in terminal or by adding a new mapping in `tmux`.

# Expected behavior
Nothing happens. It seems the `CTRL+\` sequence is completely ignored by Windows Terminal. I've tried it in Powershell and in WSL and same issue. I've also tried with another terminal emulator (kitty) and `CTRL+\` works fine there.

I found related issues: #1288, #521 but they are marked resolved and I'm still getting the bug. Any idea why?

# Actual behavior
The `CTRL+\` sequence should work.

**Comment by zadjii-msft** [comment]
What keyboard layout do you use? Is it one that has an AltGr key?

Can you share the output of running `showkey -a`, then pressing <kbd>ctrl+\\</kbd>? It should look something like:
![image](https://user-images.githubusercontent.com/18356694/100876612-3b8b5a00-346d-11eb-8ef9-7e98cd4f3b98.png)

**Comment by ojroques** [comment]
I'm using the [UK extended](http://kbdlayout.info/kbdukx/) layout so yes it has an AltGr key. I did try `showkey -a` but `Ctrl-\` outputs nothing. Other Ctrl sequences works fine though.

**Comment by alex-glv** [comment]
I am having the same problem with third party Colemak layout. Can't figure out why it's not passing through. In my case, I am trying Ctrl+_ (undo in Emacs). showkey also shows no input.

**Comment by ojroques** [comment]
I'm aware that this isn't a high priority issue since very few users should be affected. But it's quite annoying to me because it prevents me to use some features of my editor of choice (the built-in terminal of Neovim).

I would like to take a look at the code and maybe try to find the issue myself. @zadjii-msft could you tell me where I should start looking?

**Comment by DHowett** [comment]
The best place to start looking will be in Terminal.cpp's `SendKeyEvent` and `SendCharEvent`.

If you'd like to give us a quick debug trace, would you mind following the "input log" trace steps in [this issue comment](https://github.com/microsoft/terminal/issues/7067#issuecomment-664670923)? Type out a couple <kbd>Ctrl+\</kbd> sequences and send a screenshot (so we can see which records are red and which are not)

Thanks!

**Comment by ojroques** [comment]
Thanks for the info I'll look into the files your mentioned.

Here's the trace of me doing CTRL+\ using the UK extended keyboard layout:
![image](https://user-images.githubusercontent.com/23409060/102464990-62d94e00-404d-11eb-84bb-8305bae5ddf3.png)

There seems to be no trace of the release of \\. When I change my keyboard layout to US, where the sequence works fine, I get this:
![image](https://user-images.githubusercontent.com/23409060/102466080-c9ab3700-404e-11eb-9062-efc91f4cdf8c.png)

**Comment by ojroques** [comment]
I've checked the source and it seems it's due to this particular line:
https://github.com/microsoft/terminal/blob/b8e6b8e27c2c902ce18798904ce09ff32d2f68fd/src/cascadia/TerminalCore/Terminal.cpp#L763 (related doc: https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-tounicodeex).

That function returns 0 meaning that the given vkey (here 220) has no associated Unicode character. And this is what we see in the screenshot when we use UK extended layout:
![image](https://user-images.githubusercontent.com/23409060/102502655-2f171c00-407f-11eb-9ed4-d3109b1f0bc8.png)

However what I don't understand is that when I switch to the standard UK layout, I see the same vkey but a different value for the unicode character:
![image](https://user-images.githubusercontent.com/23409060/102502853-671e5f00-407f-11eb-9679-381bdbcb903f.png)

I'd expect that for the same vkey + scan code we would get the same Unicode character... what is happening there?

**Comment by zadjii-msft** [comment]
You know, that's a good question. I don't know what's happening there. Maybe @lhecker might?

**Comment by lhecker** [comment]
I'll look into it. 👍

**Comment by lhecker** [comment]
@ojroques We use `ToUnicodeEx` to determine whether a given key event will likely produce a character (like "a") as opposed to a control character or similar (like pressing an arrow key). This is important as the operating system knows more about the current state of the keyboard than our application and will help us to support dead keys (`´+e = é` for instance).

@zadjii-msft This issue occurs, because some non-US keyboard layouts apparently do not contain Ctrl-key mappings at all:
![wt-issue-8458](https://user-images.githubusercontent.com/2256941/105634204-350dd300-5e5d-11eb-8bcb-03cf50f38b5b.png)

In fact it turns out that the affected keyboards layouts don't contain _any_ mappings for Ctrl-key combinations. It appears as if Windows simply defaults a combination of Ctrl and any A-Z key to ^A-^Z, if the current keyboard layout has no mappings itself.
AFAICS the solution is to Ctrl+key combinations in `TerminalInput::HandleKey` even if `GetCharData` is zero. PR incoming.

**Comment by DHowett** [comment]
Triaged into 1.7 since we've got our best people looking at it. :smile:

**Comment by XL64** [comment]
I have similar issue with french layout and ctrl+_ not correctly passed to the linux applications but as it can be reproduced within any other windows terminals I tried, i guess it's more a wsl issue. I opened an issue on WSL github repo : https://github.com/microsoft/WSL/issues/11041

**Comment by j4james** [comment]
@XL64 That sounds like #11194, which should have been fixed in the v1.20 preview release. I've just done a test with a French keyboard layout in WSL, and <kbd>Ctrl</kbd>+<kbd>_</kbd> is definitely working as expected for me. If you can still reproduce the issue on version 1.20, it's best you open another issue and provide more details.

**Comment by XL64** [comment]
Indeed, it is fixed.  what is strange is that other native windows terminal are affected too.
